event-driven/src/lib.rs

```diff
@@ -23,48 +23,40 @@ pub use event_driven_macros::impl_fsm;
 /// and then causing a state transition in relation to that result. While this approach adds
 /// steps to a state machine, it does allow them to remain responsive to receiving more
 /// commands.
-///
-/// The generic types refer to:
-/// S  = State          - the state of your FSM
-/// C  = Command        - the command(s) that are able to be processed on your FSM
-/// E  = Event          - the event(s) that are emitted having performed a command
-/// SE = State Effect   - the effect handler
-pub trait Fsm<S, C, E, SE> {
+pub trait Fsm: Sized {
+    /// The state managed by the FSM
+    type S;
+    /// The command(s) that are able to be processed by the FSM
+    type C;
+    /// The event emitted having performed a command
+    type E;
+    /// The side effect handler
+    type SE;
+
     /// Given a state and command, optionally emit an event. Can perform side
     /// effects along the way. This function is generally only called from the
     /// `run` function.
-    fn for_command(s: &S, c: C, se: &mut SE) -> Option<E>;
+    fn for_command(s: &Self::S, c: Self::C, se: &mut Self::SE) -> Option<Self::E>;
 
-    /// Given a state and event, produce a transition, which could transition to
+    /// Given a state and event, modify state, which could indicate transition to
     /// the next state. No side effects are to be performed. Can be used to replay
     /// events to attain a new state i.e. the major function of event sourcing.
-    fn on_event(s: &S, e: &E) -> Option<S>;
-
-    /// Optional effect on exiting a state i.e. transitioning out of state `S` into
-    /// another.
-    fn on_exit(_s: &S, _se: &mut SE) {}
+    fn on_event(s: &mut Self::S, e: &Self::E) -> Option<Self::S>;
```

Comment:
I think this should just return a bool (or equivalent).   Note that returning None is not a guarantee that s was not replaced with a new macro state.

Reply:
Done